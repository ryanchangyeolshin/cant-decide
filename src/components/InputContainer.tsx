import React, { useContext } from 'react';
import { makeStyles, createStyles, Theme, } from '@material-ui/core/styles';
import { TextField, Button } from '@material-ui/core';
import DecisionContext from '../context/DecisionContext';

type InputContainerProps = {
  choice: string,
  handleKeyDown: React.KeyboardEventHandler<HTMLDivElement>,
  handleChoiceChange: React.ChangeEventHandler<HTMLInputElement>,
  randomizeChoice: React.MouseEventHandler<HTMLButtonElement>
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    inputContainer: {
      display: "flex",
      flexDirection: "row",
      justifyContent: "space-between",
      width: "100%",
      paddingLeft: "20px",
      paddingRight: "20px",
    },
    choiceInput: {
      width: "100%",
      margin: "40px 10px 10px 0",
      color: "white",
      "& .MuiFormLabel-root.Mui-focused": {
        color: "white"
      },
      "& .MuiOutlinedInput-root": {
        "& .MuiOutlinedInput-notchedOutline": {
          borderColor: "white"
        }
      }
    },
    input: {
      color: "white",
      outline: 1,
    },
    label: {
      color: "white",
    },
    randomize: {
      backgroundColor: "white",
      color: "black",
      marginTop: "40px",
      marginBottom: "10px",
      "&:hover": {
        backgroundColor: "yellow"
      },
      "& .MuiButton-label": {
        fontWeight: 900
      }
    }
  }),
);

const InputContainer = ({ choice, handleKeyDown, handleChoiceChange, randomizeChoice }: InputContainerProps) => {
  const { decision } = useContext(DecisionContext);
  const classes = useStyles();

  return (
    <div className={classes.inputContainer}>
      <TextField
        id="standard-basic"
        className={classes.choiceInput}
        label="Choice"
        variant="outlined"
        margin="normal"
        placeholder="Please provide your choice here..."
        InputLabelProps={{
          className: classes.label
        }}
        inputProps={{
          className: classes.input
        }}
        onChange={handleChoiceChange}
        onKeyDown={handleKeyDown}
        value={choice}
      />
      <Button
        className={classes.randomize}
        variant="contained"
        color="primary"
        disabled={decision.choices.length < 2}
        onClick={randomizeChoice}
      >
        Randomize
      </Button>
    </div>
  )
}

export default InputContainer;