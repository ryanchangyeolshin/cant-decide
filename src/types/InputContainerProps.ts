type InputContainerProps = {
  choice: string,
  handleKeyDown: React.KeyboardEventHandler<HTMLDivElement>,
  handleChoiceChange: React.ChangeEventHandler<HTMLInputElement>,
  randomizeChoice: React.MouseEventHandler<HTMLButtonElement>
}

export default InputContainerProps;