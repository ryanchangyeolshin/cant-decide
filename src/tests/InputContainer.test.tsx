import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import InputContainer from '../components/InputContainer';
import InputContainerProps from '../types/InputContainerProps';

afterEach(cleanup);

describe("This suit is to test the InputContainer component", () => {
  const props: InputContainerProps = {
    choice: "Taco Bell",
    handleKeyDown: jest.fn(() => "keydown"),
    handleChoiceChange: jest.fn(() => "handle choice change"),
    randomizeChoice: jest.fn(() => "randomize choice"),
  };

  test('Snapshot of ChoiceCardContainer', () => {
    const { asFragment } = render(
      <InputContainer
        choice={props.choice}
        handleKeyDown={props.handleKeyDown}
        handleChoiceChange={props.handleChoiceChange}
        randomizeChoice={props.randomizeChoice}
      />
    );
    expect(asFragment()).toMatchSnapshot()
  });

  test('that the choice text is displayed on the input field', async () => {
    const { container } = await render(
      <InputContainer
        choice={props.choice}
        handleKeyDown={props.handleKeyDown}
        handleChoiceChange={props.handleChoiceChange}
        randomizeChoice={props.randomizeChoice}
      />
    );

    const input = container.querySelector("#standard-basic") as HTMLInputElement;
    expect(input.value).toBe(props.choice);
  });

  test('that the handleKeyDown and handleChoiceChange functions are called once the user presses a key on the input field', async () => {
    const { container } = await render(
      <InputContainer
        choice={props.choice}
        handleKeyDown={props.handleKeyDown}
        handleChoiceChange={props.handleChoiceChange}
        randomizeChoice={props.randomizeChoice}
      />
    );
    const input = container.querySelector("#standard-basic") as HTMLInputElement;
    fireEvent.change(input, { target: { value: 'A' } });
    fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' });
    expect(props.handleChoiceChange).toBeCalled();
    expect(props.handleKeyDown).toBeCalled();
  });

  test('that the RANDOMIZE button is rendered', async () => {
    const { container } = await render(
      <InputContainer
        choice={props.choice}
        handleKeyDown={props.handleKeyDown}
        handleChoiceChange={props.handleChoiceChange}
        randomizeChoice={props.randomizeChoice}
      />
    );
    const randomizeButton = container.querySelector('button') as HTMLElement;
    expect(randomizeButton.textContent).toBe("Randomize");
  });
});
