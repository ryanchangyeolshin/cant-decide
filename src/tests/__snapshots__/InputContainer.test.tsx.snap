// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`This suit is to test the InputContainer component Snapshot of ChoiceCardContainer 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-inputContainer-1"
  >
    <div
      class="MuiFormControl-root MuiTextField-root makeStyles-choiceInput-2 MuiFormControl-marginNormal"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root makeStyles-label-4 MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
        data-shrink="true"
        for="standard-basic"
        id="standard-basic-label"
      >
        Choice
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input makeStyles-input-3"
          id="standard-basic"
          placeholder="Please provide your choice here..."
          type="text"
          value="Taco Bell"
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-6 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-8 PrivateNotchedOutline-legendNotched-9"
          >
            <span>
              Choice
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-randomize-5 MuiButton-containedPrimary Mui-disabled Mui-disabled"
      data-testid="randomize-button"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Randomize
      </span>
    </button>
  </div>
</DocumentFragment>
`;
